### PR TITLE
Squire client integration

### PR DESCRIPTION
**Pull request by TylerBloom**
NOTE: This PR is not ready to merge yet.

This PR moves the tournament management from a `DashMap` to a `SquireClient`. This will allow for asynchronous updates to be sent to and received from the backend server.

To finalize this PR, two major questions need to be answered:
 1) Can the ownership of persisted files be moved to the Rust side? This would include moving startup logic from the C++ side to Rust side that would check various locations for persisted files
 2) How should account and session data be persisted?

A final note, any existing tournament data files will no longer work. The client stores `TournamentManager` data, not just `Tournament` data. This should not be an issue as we don't have an existing user base.

### CODE REVIEW

**Comment by djpiper28** [comment]
is this ready to be merged?

**Comment by TylerBloom** [comment]
> is this ready to be merged?

No, the two questions in the initial PR description need to be answered.

**Comment by djpiper28** [comment]
> 1. Can the ownership of persisted files be moved to the Rust side? This would include moving startup logic from the C++ side to Rust side that would check various locations for persisted files 

This sounds logical. Would this include the configuration file

> 2.How should account and session data be persisted?

I was thinking of storing the session information and account info in the configuration file whether this is on the rust side or not does not matter too much.

**Comment by TylerBloom** [comment]
What all is contained in the config file?

**Comment by djpiper28** [comment]
> What all is contained in the config file?

default settings for tournaments, where to save tournaments, recently opened tournaments, version info

```json
{
    "default-settings": {
        "match-size": 4,
        "max-deck-count": 2,
        "min-deck-count": 1,
        "points-bye": 3,
        "points-draw": 1,
        "points-loss": 0,
        "points-win": 5,
        "type": 0
    },
    "recently-opened": [

    ],
    "remember-user": false,
    "report-crashes": false,
    "tourn-save-path": "tourns/",
    "user": {
        "user-name": "",
        "user-token": ""
    },
    "version": "0.1 (130a36d@main)"
}
```

**Comment by djpiper28** [comment]
> Ah, ok. All of that can be stored on the Rust side. If there are any UI-specific configs (e.g. color schemes), those configs can be stored in a separate file and the Rust config can point the UI to that file location. This should avoid a messy situation where we need to pass config across the language barrier.
> 
> Nit: Version can probably be a baked-in value. At least in Rust, there are libraries that will pull the version number from your `Cargo.toml` (and optionally append build numbers/git commit hashes).

I wanted to store the version in the config to warn when it downgrades or upgrades

**Comment by TylerBloom** [comment]
> I wanted to store the version in the config to warn when it downgrades or upgrades

Can't that be part of an upgrade manifest? The current, running desktop is aware of its version, it fetches a different version and can warn the user of any version changes. We want to prevent a user from being able to skip a required upgrade by simply changing the config file.

**Comment by djpiper28** [comment]
> > I wanted to store the version in the config to warn when it downgrades or upgrades
> 
> Can't that be part of an upgrade manifest? The current, running desktop is aware of its version, it fetches a different version and can warn the user of any version changes. We want to prevent a user from being able to skip a required upgrade by simply changing the config file.

I think storing the version in the config then notfying the user is far easier than dealing with upgrade manifests

**Comment by djpiper28** [comment]
> > I think storing the version in the config then notfying the user is far easier than dealing with upgrade manifests
> 
> But won't we want upgrade manifests to distribute changelogs and such?

we will want them, but the issue with manifests is how do you know when it was updated? this is why the version is stored in the config so that it is known when the software was updated.

**Comment by TylerBloom** [comment]
I suppose having version info in the config is ok for now. We can always move them into manifests later (once we get a better idea of what those manifests will look like).

I think this answer most of my questions. I just have a few minor questions, and then I'll be able to finish this.

1) Where is the config file stored? Next to the executable?
2) Is the save path in the config (e.g. `"tourn-save-path": "tourns/"`) an absolute or relative path?
3) Does the Rust runtime need to expose any other methods to help the C++ side get notified when a remote update occurs?

File: squire_ffi/src/lib.rs [hunk 1]
```diff
@@ -1,175 +1,141 @@
...
-pub extern "C" fn sq_free(pointer: *mut c_void, len: usize) {
-    unsafe {
-        System.deallocate(
-            ptr::NonNull::new(pointer as *mut u8).unwrap(),
-            Layout::from_size_align(len, 1).unwrap(),
-        );
+impl SquireRuntime {
+    /// Creates the runtime.
+    pub fn new() -> Self {
+        todo!()
```
Comment (TylerBloom) [inline squire_ffi/src/lib.rs:75]: This needs to be implemented still...

File: squire_ffi/src/lib.rs [hunk 2]
```diff
@@ -1,175 +1,141 @@
...
+        self.client.import_tourn(tourn).process_blocking()
+    }
+
+    /// Removes a tournament from the runtime and returns it, if found
+    pub fn remove_tournament(&self, t_id: TournamentId) -> Result<(), ActionError> {
+        self.client
+            .remove_tourn(t_id)
+            .process_blocking()
             .ok_or_else(|| ActionError::TournamentNotFound(t_id))
+            .map(|_| ())
```
Comment (TylerBloom) [inline squire_ffi/src/lib.rs:127]: nit: This should be `drop` or some other function like `forget`.

File: squire_ffi/src/tournament.rs [hunk 3]
```diff
@@ -61,13 +63,15 @@ pub extern "C" fn tid_standings(tid: TournamentId) -> *const PlayerScore<Standar
...
 #[no_mangle]
 pub extern "C" fn tid_players(tid: TournamentId) -> *const PlayerId {
-    match SQUIRE_RUNTIME
-        .get()
-        .unwrap()
-        .tournament_query(tid, |t| unsafe {
-            copy_to_system_pointer(t.player_reg.players.keys().cloned().map(Into::into))
-        }) {
-        Ok(data) => data,
+    match CLIENT.get().unwrap().tournament_query(tid, |t| {
```
Comment (TylerBloom) [inline squire_ffi/src/tournament.rs:66]: There should be a `query_players` method that should be used here.

File: squire_ffi/src/tournament.rs [hunk 4]
```diff
@@ -81,13 +85,15 @@ pub extern "C" fn tid_players(tid: TournamentId) -> *const PlayerId {
...
 pub extern "C" fn tid_rounds(tid: TournamentId) -> *const RoundId {
-    match SQUIRE_RUNTIME
-        .get()
-        .unwrap()
-        .tournament_query(tid, |t| unsafe {
-            copy_to_system_pointer(t.round_reg.num_and_id.iter_right().cloned().map(Into::into))
-        }) {
-        Ok(data) => data,
+    match CLIENT.get().unwrap().tournament_query(tid, |t| {
+        t.round_reg
```
Comment (TylerBloom) [inline squire_ffi/src/tournament.rs:89]: Same thing, use the `query_rounds` method.

File: squire_ffi/src/tournament.rs [hunk 5]
```diff
@@ -267,130 +273,93 @@ pub extern "C" fn tid_update_settings(
     // Sort input strings out
     let format = String::from(unsafe { CStr::from_ptr(__format).to_str().unwrap().to_string() });
 
-    let rt = SQUIRE_RUNTIME.get().unwrap();
+    let client = CLIENT.get().unwrap();
+    let curr_max = client
+        .tournament_query(tid, |t| t.settings.max_deck_count)
+        .unwrap();
+    let (deck_op_one, deck_op_two) = if min_deck_count > curr_max {
```
Comment (TylerBloom) [inline squire_ffi/src/tournament.rs:280]: Replace this if-else with something like this:
```rust
let mut deck_op_one = /* Update min */;
let mut deck_op_two = /* Update max */;
if min_deck_count > curr_max {
  std::mem::swap(&mut deck_op_one, &mut deck_op_two);
}
*/

File: squire_ffi/src/tournament.rs [hunk 6]
```diff
@@ -267,130 +273,93 @@ pub extern "C" fn tid_update_settings(
...
+        (
+            AdminOp::UpdateTournSetting(TournamentSetting::GeneralSetting(
```
Comment (TylerBloom) [inline squire_ffi/src/tournament.rs:282]: All operations should use `Into` on their inner settings to reduce boilerplate.

**Comment by djpiper28** [comment]
> I suppose having version info in the config is ok for now. We can always move them into manifests later (once we get a better idea of what those manifests will look like).
> 
> I think this answer most of my questions. I just have a few minor questions, and then I'll be able to finish this.
> 
>     1. Where is the config file stored? Next to the executable?
> 
>     2. Is the save path in the config (e.g. `"tourn-save-path": "tourns/"`) an absolute or relative path?
> 
>     3. Does the Rust runtime need to expose any other methods to help the C++ side get notified when a remote update occurs?

1. the config file is stored in the working directory, I am working on a wrapper to set the working directory to a suitable place on launch
2. absolute path, it just calls `fopen(tourn_save_path + "/" + name)` basically
3. some sort of method to poll such that the ui can call it periodically then get the update type

File: squire_ffi/src/lib.rs [hunk 7]
```diff
@@ -1,175 +1,141 @@
...
-    TournamentNotFound(TournamentId),
-    /// The given round could not be found
-    RoundNotFound(TournamentId, RoundId),
-    /// The given player could not be found
-    PlayerNotFound(TournamentId, PlayerId),
-    /// An wrap for a tournament error
-    OperationError(TournamentId, TournamentError),
-}
+    /// Checks to see if an update to a tournament has been sent from the backend
+    pub fn poll_remote_update(&mut self, t_id: TournamentId) -> bool {
```
Comment (TylerBloom) [inline squire_ffi/src/lib.rs:79]: This needs to be exposed as a `TournamentId` method.

**Pull request merged**